engine/api/services/http.go

```diff
@@ -20,6 +22,67 @@ import (
 // HTTPClient will be set to a default httpclient if not set
 var HTTPClient sdk.HTTPClient
 
+// DoMultiPartRequest performs an http request on a service with multipart  tar file + json field
+func DoMultiPartRequest(ctx context.Context, srvs []sdk.Service, method, path string, buf *bytes.Buffer, in interface{}, out interface{}, mods ...sdk.RequestModifier) (int, error) {
+
+	body := &bytes.Buffer{}
+	writer := multipart.NewWriter(body)
+
+	// Create tar part
+	tarPartMH := make(textproto.MIMEHeader)
+	tarPartMH.Set("Content-Type", "application/tar")
+	tarPartMH.Set("Content-Disposition", "form-data; name=\"dataFiles\"; filename=\"data.tar\"")
+	dataPart, err := writer.CreatePart(tarPartMH)
+	if err != nil {
+		return 0, sdk.WrapError(err, "unable to create data part")
+	}
+	if _, err := io.Copy(dataPart, buf); err != nil {
+		return 0, sdk.WrapError(err, "unable to write into data part")
+	}
+
+	jsonData, errM := json.Marshal(in)
+	if errM != nil {
+		return 0, sdk.WrapError(errM, "unable to marshal data")
+	}
+	if err := writer.WriteField("dataJSON", string(jsonData)); err != nil {
+		return 0, sdk.WrapError(err, "unable to add field dataJSON")
+	}
+
+	// Close writer
+	if err := writer.Close(); err != nil {
+		return 0, sdk.WrapError(err, "unable to close writer")
+	}
+
+	mods = append(mods, sdk.SetHeader("Content-Type", writer.FormDataContentType()))
+	var lastErr error
+	var lastCode int
+	var attempt int
+	for {
+		attempt++
+		for i := range srvs {
+			srv := &srvs[i]
+			res, code, err := doRequest(ctx, srv.HTTPURL, srv.Hash, method, path, body.Bytes(), mods...)
+			if err != nil {
+				return code, sdk.WrapError(err, "Unable to perform request on service %s (%s)", srv.Name, srv.Type)
+			}
+			if out != nil {
+				if err := json.Unmarshal(res, out); err != nil {
+					return code, sdk.WrapError(err, "Unable to marshal output")
+				}
+			}
+			if err == nil {
+				return code, nil
+			}
+			lastErr = err
+			lastCode = code
+		}
+		if lastErr != nil || attempt > 5 {
```

Comment:
Why this will break if lastErr != nil ? Looks like we want to ask for result to all services five time max.
But all services returns an error there will be only one attempt.

Reply:
We just want 1 result from 1 service.
If there is an erorr on 1 service, we try another 1.